### PR TITLE
feat: expose uiSchema and formContext in FormProps

### PR DESCRIPTION
**Pull request by richcooper95**
## Hey, I just made a Pull Request!

This PR exposes two extra fields in the FormProps: `uiSchema` and `formContext`.

This allows more control over how specific form fields are rendered when using the Scaffolder.

In our specific use-case, we are using the `EmbeddableWorkflow` Component introduced in https://github.com/backstage/backstage/pull/15162 to update resources, and want to to make non-editable fields read-only and disabled in the UI.

With this change, we can do the following:

```
        <EmbeddableWorkflow
          namespace={namespace}
          templateName={templateName}
          onError={onError}
          onCreate={onCreate}
          extensions={customFieldExtensions}
          initialState={initialState}
          formProps={{
            uiSchema: uiSchema,
            formContext: {
              readonlyAsDisabled: true
            }
          }}
        />
```

with `uiSchema` defined to set `myReadOnlyField` to read-only:

```
  const uiSchema = {
    myReadOnlyField: {
      "ui:readonly": true, // This makes the field read-only in the UI
    },
  };
```

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
~- [ ] Added or updated documentation~
~- [ ] Tests for new functionality and regression tests for bug fixes~
~- [ ] Screenshots attached (for UI changes)~
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
Hey @richcooper95 :wave: thanks for this PR!

I think that we're going to want to add support for merging the properties with the current `ui:schema` and `formContext` otherwise providing ones to the `Workflow` would overwrite the ones that are already provided: https://github.com/backstage/backstage/blob/0e692cf4227d4a03a81981979669935e945665ca/plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx#L227-L228

I think we'll need to destructure them out of `formProps` and merge them seperately.

**Comment by richcooper95** [comment]
Hi @benjdlambert - apologies for the long delay in getting back to this!

We're looking to pick this PR up again - any chance you could please reopen it? 🙏

Thanks for the pointer too! I've just had a look and I think what you're saying is we'd need something like this, passing each into the Stepper's `Form`:

```typescript
  var formContext = { formData: formState };
  if (props.formProps?.formContext) {
    formContext = { ...props.formProps.formContext, ...formContext };
  }

  var uiSchema = currentStep.uiSchema;
  if (props.formProps?.uiSchema) {
    uiSchema = { ...props.formProps.uiSchema, ...uiSchema };
  }
```

Is that what you had in mind?

**Comment by benjdlambert** [comment]
@richcooper95 sorry been out on holidays and just catching up! Yes, indeed! Something like that, but you can do it inline in the `props` of the `Component`.

```tsx
<Stepper
  ...
  uiSchema={{ ...uiSchema, ...formProps?.uiSchema }}
 ```

**Comment by richcooper95** [comment]
Thanks @benjdlambert, and no worries at all! I've just made the changes - I _think_ it's right that the new props are overwritten by the Stepper values, i.e.:

- `{ formData: formState }` takes precedence over the new `formProps.formContext` (so the right state is rendered)
- Values in `currentStep.uiSchema` take precedence over the new `formProps.uiSchema` (though I'm less sure about this one - my thinking is that the `formProps.uiSchema` should only override schema fields which aren't set in the `currentStep.uiSchema`)

Does this seem correct to you? (It seems to work for what we wanted to achieve when testing locally, but I might be missing something!)

File: .changeset/flat-plums-grow.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-scaffolder-react': patch
+---
+
+Add ability to customise form fields in the UI by exposing uiSchema and formContext in FormProps
```
Comment (benjdlambert) [inline .changeset/flat-plums-grow.md:5]:
```suggestion
Add ability to customise form fields in the UI by exposing `uiSchema` and `formContext` in `FormProps`
```

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Think this is looking pretty good, Would like to have some tests for this as well. You can test it by providing some overrides for the `formContext` and the `uiSchema` and assert that they are in the rendered form. Like providing an override for a field or something and checking that it's not provided in the template but is actually rendered by default.

Just a small thing in the changeset too :pray:

File: plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx [hunk 2]
```diff
@@ -51,6 +51,7 @@ import { ErrorListTemplate } from './ErrorListTemplate';
 import { makeStyles } from '@material-ui/core/styles';
 import { PasswordWidget } from '../PasswordWidget/PasswordWidget';
 import ajvErrors from 'ajv-errors';
+import merge from 'lodash.merge';
```
Comment (benjdlambert) [inline plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:54]: `lodash` is already installed, so you should be able to do

```ts
import { merge } from 'lodash'
```

**Review by benjdlambert — COMMENTED** [review]
Just a comment about re-using lodash

**Comment by richcooper95** [comment]
> Just a comment about re-using lodash

Good shout - I think this was causing some CI failures during `yarn tsc` too. Made the change - thanks! 👍

**Review by benjdlambert — APPROVED** [review]
Nice! Let's go with this :tada:

**Comment by richcooper95** [comment]
Thanks for merging, and for your help getting this done! 🙏

**Pull request merged**